File: KingdomDivision/CPP/KingdomDivision.cpp
#include <iostream>
#include <string>
#include <vector>
#include <map>

using namespace std;

#define mod 1000000007;

enum
{
    safe,
    unsafe
};

void getDivisions(vector<vector<long> *> *roadsMap, vector<bool> *visited, vector<vector<long> *> *divisions, int index = 0)
{
    (*visited)[index] = true;
    long sf = 1;
    long usf = 1;

    for (long i : *(*roadsMap)[index])
    {
        if ((*visited)[i])
        {
            continue;
        }

        getDivisions(roadsMap, visited, divisions, i);

        usf *= (*(*divisions)[i])[safe];
        usf %= mod;

        sf *= ((*(*divisions)[i])[safe] * 2) + (*(*divisions)[i])[unsafe];
        sf %= mod;
    }

    sf -= usf;
    sf += mod;
    sf %= mod;
    (*(*divisions)[index])[safe] = sf;
    (*(*divisions)[index])[unsafe] = usf;
}

int kingdomDivision(int n, vector<vector<int>> &roads)
{
    vector<vector<long> *> *roadsMap = new vector<vector<long> *>(n);

    vector<vector<long> *>::iterator it;
    int i = 0;
    for (it = roadsMap->begin(); it != roadsMap->end(); it++, i++)
    {
        (*roadsMap)[i] = new vector<long>();
    }

    for (vector<int> road : roads)
    {
        road[0]--;
        road[1]--;
        (*roadsMap)[road[0]]->push_back(road[1]);
        (*roadsMap)[road[1]]->push_back(road[0]);
    }

    vector<vector<long> *> *divisions = new vector<vector<long> *>(n, new vector<long>(2));

    getDivisions(roadsMap, new vector<bool>(n), divisions);

    return (2 * (*(*divisions)[0])[safe]) % mod;
}

int main()
{
    int n;
    cin >> n;

    vector<vector<int>> roads(n - 1);
    for (int i = 0; i < n - 1; i++)
    {
        roads[i] = vector<int>(2);

        for (int j = 0; j < 2; j++)
        {
            cin >> roads[i][j];
        }
    }

    int result = kingdomDivision(n, roads);

    cout << result << "\n";

    return 0;
}

File: KingdomDivision/Python/KingdomDivision.py
#!/bin/python3

MOD = 1000000007
SAFE = 0
UNSAFE = 1


def get_divisions(roads_map, visited, divisions, index=0):
    visited[index] = True
    sf = 1
    usf = 1

    for i in roads_map[index]:
        if visited[i]:
            continue

        get_divisions(roads_map, visited, divisions, i)

        usf = usf * divisions[i][SAFE]
        usf = usf % MOD

        sf = sf * ((divisions[i][SAFE] * 2) + divisions[i][UNSAFE])
        sf = sf % MOD

    sf = sf - usf
    sf = sf + MOD
    sf = sf % MOD
    divisions[index][SAFE] = sf
    divisions[index][UNSAFE] = usf


def kingdom_division(n, roads):
    roads_map = [None] * n
    divisions = [None] * n

    for i in range(n):
        roads_map[i] = []
        divisions[i] = [0] * 2

    for road in roads:
        road[0] -= 1
        road[1] -= 1
        roads_map[road[0]].append(road[1])
        roads_map[road[1]].append(road[0])

    visited = [False] * n

    get_divisions(roads_map, visited, divisions)

    return (2 * divisions[0][SAFE]) % MOD


if __name__ == '__main__':
    n = int(input())

    roads = []

    for _ in range(n-1):
        roads.append(list(map(int, input().rstrip().split())))

    result = kingdom_division(n, roads)

    print(str(result) + '\n')
